server/api/orphaned-record/orphaned-record.controller.ts

```diff
@@ -100,81 +103,102 @@ class OrphanedRecordController {
+  async resolveOrphanedRecordWithAdd(req: ApiRequest<OrphanedRecordResolveParam, RosterEntryData>, res: Response) {
+    const compositeId = req.params.orphanId;
+    assertRequestBody(req, [
+      'edipi',
+      'unit',
+    ]);
+    const entryData = req.body;
+
     const orphanedRecords = await OrphanedRecord.find({
       where: {
-        compositeId: req.params.orphanId,
+        compositeId,
         deletedOn: null,
       },
     });
 
     if (orphanedRecords.length === 0) {
-      throw new BadRequestError(`Unable to locate orphaned record with id: ${req.params.orphanId}`);
-    }
-    if (orphanedRecords.length > 1) {
-      throw new BadRequestError(`Encountered Multiple Orphaned Records: ${req.params.orphanId}`);
+      throw new BadRequestError(`Unable to locate orphaned record with id: ${compositeId}`);
     }
 
-    const orphanedRecord = orphanedRecords[0];
-    const documentId = orphanedRecord.documentId;
-    let rosterHistory: RosterHistory[] = [];
+    // We have to modify the roster outside of the transaction, since the orphaned record resolve process
+    // depends on having the latest roster history from this addition.
+    const entry = await getManager().transaction(async manager => {
+      return addRosterEntry(req.appOrg!, req.appUserRole!.role, entryData, manager);
+    });
 
-    if (req.body.unit) {
-      rosterHistory = await getRosterHistoryForOrphanedRecord(orphanedRecord, req.body.unit);
-    }
+    try {
+      await getManager().transaction(async manager => {
+        for (const orphanedRecord of orphanedRecords) {
+          await orphanedRecord.resolve(entry, manager);
+        }
+      });
+    } catch (err) {
+      // If something went wrong we have to rollback the new entry manually.
+      await entry.remove();
 
-    let newRosterEntry: Roster | undefined;
-    if (!rosterHistory.length) {
-      // Add a new roster entry since the orphaned record
-      // doesn't correspond to an existing roster entry
-      newRosterEntry = await addRosterEntry(req.appOrg!, req.appUserRole!.role, req.body);
-      rosterHistory = await getRosterHistoryForOrphanedRecord(orphanedRecord, newRosterEntry.unit.id);
+      throw err;
     }
 
-    if (!rosterHistory.length) {
-      throw new InternalServerError('Unable to locate RosterHistory record.');
+    res.status(200).send();
+  }
+
+  async resolveOrphanedRecordWithEdit(req: ApiRequest<OrphanedRecordResolveParam, ResolveWithEditBody>, res: Response) {
+    const compositeId = req.params.orphanId;
+    assertRequestBody(req, [
+      'edipi',
+      'unit',
+    ]);
+    const { id: entryId, ...entryData } = req.body;
+
+    const orphanedRecords = await OrphanedRecord.find({
+      where: {
+        compositeId,
+        deletedOn: null,
+      },
+    });
+
+    if (orphanedRecords.length === 0) {
+      throw new BadRequestError(`Unable to locate orphaned record with id: ${compositeId}`);
```

Comment:
I opted for simplifying to this, so that we can leave the error handling up to the caller.

```typescript
    const orphanedRecords = await getOrphanedRecordsForResolve(compositeId);
    if (orphanedRecords.length === 0) {
      throw new OrphanedRecordsNotFoundError(compositeId);
    }
```

At some point, if we ever get time for tech debt, we should go through and create error types like this for anything that happens in more than one place.